crates/hc_sandbox/src/save.rs

```diff
@@ -2,125 +2,421 @@
 //!
 //! This module gives basic helpers to save / load your sandboxes
 //! to / from a `.hc` file.
-//! This is very much WIP and subject to change.
 
+use crate::cmds::Existing;
+use holochain_conductor_api::conductor::paths::ConfigRootPath;
+use holochain_conductor_api::{AdminInterfaceConfig, InterfaceDriver};
+use holochain_conductor_config::config::read_config;
+use std::io::Write;
 use std::path::{Path, PathBuf};
 use std::sync::OnceLock;
 
-use anyhow::Context;
-use holochain_conductor_api::conductor::paths::{ConfigFilePath, ConfigRootPath};
+/// Representation of a loaded `.hc` file
+#[derive(Debug)]
+pub struct HcFile {
+    /// Path to the directory containing the `.hc` file.
+    pub dir: PathBuf,
+    /// Vec of results after trying to read each path in the `.hc` file.
+    pub existing_all: Vec<Result<ConfigRootPath, ConfigRootPath>>,
+}
 
-/// Save all sandboxes to the `.hc` file in the `hc_dir` directory.
-pub fn save(mut hc_dir: PathBuf, paths: Vec<ConfigRootPath>) -> anyhow::Result<()> {
-    use std::io::Write;
-    std::fs::create_dir_all(&hc_dir)?;
-    hc_dir.push(".hc");
-    let mut file = std::fs::OpenOptions::new()
-        .append(true)
-        .create(true)
-        .open(hc_dir)?;
+impl HcFile {
+    /// Return only valid existing paths
+    pub fn valid_paths(&self) -> Vec<ConfigRootPath> {
+        self.existing_all.iter().flatten().cloned().collect()
+    }
 
-    for path in paths {
-        writeln!(file, "{}", path.display())?;
+    /// Return only invalid existing paths
+    pub fn invalid_paths(&self) -> Vec<ConfigRootPath> {
+        self.existing_all
+            .iter()
+            .filter_map(|result| result.clone().err())
+            .collect()
     }
-    Ok(())
-}
 
-/// Remove sandboxes by their index in the file.
-/// You can get the index by calling [`load`].
-/// If no sandboxes are passed in then all are deleted.
-/// If all sandboxes are deleted the `.hc` file will be removed.
-pub fn clean(mut hc_dir: PathBuf, sandboxes: Vec<usize>) -> anyhow::Result<()> {
```

Comment:
moved as a method of HcFile and renamed to `remove` and changed input to use the `Existing` struct instead of a list of indices.